crates/bevy_pbr/src/meshlet/visibility_buffer_software_raster.wgsl

```diff
@@ -0,0 +1,195 @@
+
+var<workgroup> viewport_vertices: array<vec3f, 64>;
+
+@compute
+@workgroup_size(64, 1, 1) // 64 threads per workgroup, 1 vertex/triangle per thread, 1 cluster per workgroup
+fn rasterize_cluster(
+    @builtin(workgroup_id) workgroup_id: vec3<u32>,
+    @builtin(local_invocation_index) local_invocation_index: u32,
+#ifdef MESHLET_2D_DISPATCH
+    @builtin(num_workgroups) num_workgroups: vec3<u32>,
+#endif
+) {
+    var workgroup_id_1d = workgroup_id.x;
+
+#ifdef MESHLET_2D_DISPATCH
+    workgroup_id_1d += workgroup_id.y * num_workgroups.x;
+    if workgroup_id_1d >= meshlet_software_raster_cluster_count { return; }
+#endif
+
+    let cluster_id = meshlet_raster_clusters[workgroup_id_1d];
+    let meshlet_id = meshlet_cluster_meshlet_ids[cluster_id];
+    let meshlet = meshlets[meshlet_id];
+
+    // Load and project 1 vertex per thread
+    let vertex_id = local_invocation_index;
+    if vertex_id < meshlet.vertex_count {
+        let meshlet_vertex_id = meshlet_vertex_ids[meshlet.start_vertex_id + vertex_id];
+        let vertex = unpack_meshlet_vertex(meshlet_vertex_data[meshlet_vertex_id]);
+
+        // Project vertex to viewport space
+        let instance_id = meshlet_cluster_instance_ids[cluster_id];
+        let instance_uniform = meshlet_instance_uniforms[instance_id];
+        let world_from_local = affine3_to_square(instance_uniform.world_from_local);
+        let world_position = mesh_position_local_to_world(world_from_local, vec4(vertex.position, 1.0));
+        var clip_position = view.clip_from_world * vec4(world_position.xyz, 1.0);
+        var ndc_position = clip_position.xyz / clip_position.w;
+#ifdef DEPTH_CLAMP_ORTHO
+        ndc_position.z = 1.0 / clip_position.z;
+#endif
+        let viewport_position_xy = ndc_to_uv(ndc_position.xy) * view.viewport.zw;
+
+        // Write vertex to workgroup shared memory
+        viewport_vertices[vertex_id] = vec3(viewport_position_xy, ndc_position.z);
+    }
+
+    workgroupBarrier();
+
+    // Load 1 triangle's worth of vertex data per thread
+    let triangle_id = local_invocation_index;
+    if triangle_id >= meshlet.triangle_count { return; }
+    let index_ids = meshlet.start_index_id + (triangle_id * 3u) + vec3(0u, 1u, 2u);
+    let vertex_ids = vec3(get_meshlet_index(index_ids[0]), get_meshlet_index(index_ids[1]), get_meshlet_index(index_ids[2]));
+    let vertex_0 = viewport_vertices[vertex_ids[2]];
+    let vertex_1 = viewport_vertices[vertex_ids[1]];
+    let vertex_2 = viewport_vertices[vertex_ids[0]];
+    let packed_ids = (cluster_id << 6u) | triangle_id;
+
+    // Compute triangle bounding box
+    let min_x = u32(min3(vertex_0.x, vertex_1.x, vertex_2.x));
+    let min_y = u32(min3(vertex_0.y, vertex_1.y, vertex_2.y));
+    var max_x = u32(ceil(max3(vertex_0.x, vertex_1.x, vertex_2.x)));
+    var max_y = u32(ceil(max3(vertex_0.y, vertex_1.y, vertex_2.y)));
+    max_x = min(max_x, u32(view.viewport.z) - 1u);
+    max_y = min(max_y, u32(view.viewport.w) - 1u);
```

Comment:
do you also need to `min_x = max(0, min_x)` etc?

Reply:
Iirc the u32(foo) should turn all negatives into 0.